### PR TITLE
chore: Migrate `runtime-image` to use `dockerode`

### PR DESCRIPTION
**Pull request by DominicKramer**

### CODE REVIEW

File: runtime-image/test/test.ts [hunk 1]
```diff
@@ -1,21 +1,32 @@
...
-import {exec} from 'child_process';
-import * as uuid from 'node-uuid';
+import * as Docker from 'dockerode';
+import * as path from 'path';
 import * as request from 'request';
-import * as util from 'util';
+
+const tar: {pack: (dir: string) => any} = require('tar-fs');
+
+const RUN_TIMEOUT = 3000;
```
Comment (jinwoo) [inline runtime-image/test/test.ts]: Would `RUN_TIMEOUT_MS` be better?

File: runtime-image/test/test.ts [hunk 2]
```diff
@@ -1,21 +1,32 @@
...
+const BUILD_TIMEOUT = 5 * 60 * 1000;
```
Comment (jinwoo) [inline runtime-image/test/test.ts]: Similarly `BUILD_TIMEOUT_SEC` ?

File: runtime-image/test/test.ts [hunk 3]
```diff
@@ -1,21 +1,32 @@
...
+const PORT = 8080;
+
+let host = 'localhost';
```
Comment (jinwoo) [inline runtime-image/test/test.ts]: I'll combine these into one expression and use `const`.

```typescript
const host = process.env.DOCKER_HOST ?
    process.env.DOCKER_HOST.split('//')[1].split(':')[0] :
    'localhost';
```

File: runtime-image/test/test.ts [hunk 4]
```diff
@@ -139,55 +150,86 @@ const CONFIGURATIONS: TestConfig[] = [
...
+  before(async function() {
+    this.timeout(BUILD_TIMEOUT);
+    await buildDocker(ROOT, 'test/nodejs');
+    const dir = path.join(ROOT, 'test', 'definitions', 'base-install-node');
+    await buildDocker(dir, 'test/definitions/base-install-node');
+  });
+
+  CONFIGURATIONS.forEach(config => {
+    describe(`Image ${config.tag}`, () => {
+      let container: Docker.Container|undefined = undefined;
```
Comment (jinwoo) [inline runtime-image/test/test.ts]: Nit. It'll be `undefined` if you don't assign anything.

File: runtime-image/test/test.ts [hunk 5]
```diff
@@ -139,55 +150,86 @@ const CONFIGURATIONS: TestConfig[] = [
...
+      function onFinished(err2: Error, output: any) {
+        if (err2) {
+          return reject(err2);
+        }
+
+        resolve(output);
+      }
+
+      function onProgress(event: any) {
+        log(event.stream);
```
Comment (jinwoo) [inline runtime-image/test/test.ts]: Curious. Aren't these logged just as `Object [object]` by the `log()` function?

Comment (DominicKramer) [reply runtime-image/test/test.ts]: `event.stream` etc. are all strings.

File: runtime-image/test/test.ts (hunk 4 above)
Comment (DominicKramer) [reply runtime-image/test/test.ts]: Done

File: runtime-image/test/test.ts (hunk 3 above)
Comment (DominicKramer) [reply runtime-image/test/test.ts]: Done

File: runtime-image/test/test.ts (hunk 2 above)
Comment (DominicKramer) [reply runtime-image/test/test.ts]: Done.

File: runtime-image/test/test.ts (hunk 1 above)
Comment (DominicKramer) [reply runtime-image/test/test.ts]: Yes, it sounds better.  I changed it.

**Pull request merged**